File: src/models/works_model.py
from PyQt6.QtCore import (
    Qt,
    QAbstractListModel,
    QModelIndex,
    pyqtProperty
)


class WorksModel(QAbstractListModel):

    IdRole = Qt.ItemDataRole.UserRole + 1
    ObjectIdRole = Qt.ItemDataRole.UserRole + 2
    ServiceIdRole = Qt.ItemDataRole.UserRole + 3
    SubobjectNameRole = Qt.ItemDataRole.UserRole + 4
    NameRole = Qt.ItemDataRole.UserRole + 5
    PriceRole = Qt.ItemDataRole.UserRole + 6
    UnitRole = Qt.ItemDataRole.UserRole + 7
    QuantityRole = Qt.ItemDataRole.UserRole + 8
    CreatedAtRole = Qt.ItemDataRole.UserRole + 9
    UpdatedAtRole = Qt.ItemDataRole.UserRole + 10
    StartOrderAtRole = Qt.ItemDataRole.UserRole + 11

    def __init__(self):
        super().__init__()
        self._items = []

    def updateModel(self, items):
        self.beginResetModel()
        self._items = items
        self.endResetModel()

    def clearModel(self):
        self.beginResetModel()
        self._items = []
        self.endResetModel()

    def addItem(self, item):
        if not item:
            return False

        self.beginInsertRows(QModelIndex(), 0, 0)
        self._items.insert(0, item)
        self.endInsertRows()
        return True

    def updateItem(self, item):
        work_id = item.get("id")
        if not work_id:
            return False

        for index, existing in enumerate(self._items):
            if existing["id"] != work_id:
                continue

            updated = existing.copy()
            updated.update(item)
            self._items[index] = updated
            model_index = self.index(index)
            self.dataChanged.emit(model_index, model_index)
            return True

        return False

    def removeItem(self, work_id):
        if not work_id:
            return False

        for index, item in enumerate(self._items):
            if item["id"] != work_id:
                continue

            self.beginRemoveRows(QModelIndex(), index, index)
            del self._items[index]
            self.endRemoveRows()
            return True

        return False

    def rowCount(self, parent=QModelIndex()):
        return len(self._items)

    def data(self, index, role):
        if not index.isValid():
            return None

        item = self._items[index.row()]

        if role == self.IdRole:
            return item["id"]

        if role == self.ObjectIdRole:
            return item["object_id"]

        if role == self.ServiceIdRole:
            return item["service_id"]

        if role == self.SubobjectNameRole:
            return item["subobject_name"]

        if role == self.NameRole:
            return item["name"]

        if role == self.PriceRole:
            return item["price"]

        if role == self.UnitRole:
            return item["unit"]

        if role == self.QuantityRole:
            return item["quantity"]

        if role == self.CreatedAtRole:
            return item["created_at"]

        if role == self.UpdatedAtRole:
            return item["updated_at"]

        if role == self.StartOrderAtRole:
            return item["start_order_at"]

        return None

    def roleNames(self):
        return {
            self.IdRole: b"id",
            self.ObjectIdRole: b"object_id",
            self.ServiceIdRole: b"service_id",
            self.SubobjectNameRole: b"subobject_name",
            self.NameRole: b"name",
            self.PriceRole: b"price",
            self.UnitRole: b"unit",
            self.QuantityRole: b"quantity",
            self.CreatedAtRole: b"created_at",
            self.UpdatedAtRole: b"updated_at",
            self.StartOrderAtRole: b"start_order_at",
        }

    def itemData(self, work_id):
        for item in self._items:
            if item["id"] == work_id:
                return item
        return None

    def items(self):
        return [item.copy() for item in self._items]

    @pyqtProperty(int)
    def count(self):
        return len(self._items)

File: cpp/models/worksModel.cpp
#include "worksModel.h"

#include "modelUtils.h"

namespace workorder
{

WorksModel::WorksModel(QObject *parent)
    : GenericModel<WorkItem>(parent)
{
}

void WorksModel::updateModelFromMaps(const QVariantList &items)
{
    setItems(itemsFromVariantList<WorkItem>(items, WorkItem::fromMap));
}

void WorksModel::clearModel()
{
    clear();
}

bool WorksModel::addItem(const QVariantMap &item)
{
    if(item.isEmpty())
        return false;

    beginInsertRows(QModelIndex(), 0, 0);
    m_items.insert(0, WorkItem::fromMap(item));
    endInsertRows();
    emit sizeChanged();
    return true;
}

bool WorksModel::updateItem(const QVariantMap &item)
{
    const QString workId = item.value("id").toString();
    if(workId.isEmpty())
        return false;

    for(int index = 0; index < m_items.size(); ++index)
    {
        if(m_items.at(index).id != workId)
            continue;

        WorkItem updated = m_items.at(index);
        const QVariantMap merged = updated.toMap();
        for(auto it = item.constBegin(); it != item.constEnd(); ++it)
        {
            if(it.key() == "object_id")
                updated.objectId = it.value().toString();
            else if(it.key() == "service_id")
                updated.serviceId = it.value().toString();
            else if(it.key() == "subobject_name")
                updated.subobjectName = it.value().toString();
            else if(it.key() == "name")
                updated.name = it.value().toString();
            else if(it.key() == "price")
                updated.price = it.value().toInt();
            else if(it.key() == "unit")
                updated.unit = it.value().toString();
            else if(it.key() == "quantity")
                updated.quantity = it.value().toDouble();
            else if(it.key() == "created_at")
                updated.createdAt = it.value().toLongLong();
            else if(it.key() == "updated_at")
                updated.updatedAt = it.value().toLongLong();
            else if(it.key() == "start_order_at")
                updated.startOrderAt = it.value().toLongLong();
            else if(it.key() == "coefficients")
                updated.coefficients = it.value().toString();
            else if(it.key() == "percent_sum")
                updated.percentSum = it.value().toInt();
        }
        Q_UNUSED(merged);

        m_items[index] = updated;
        const QModelIndex modelIndex = this->index(index);
        emit dataChanged(modelIndex, modelIndex);
        emit sigItemChanged(index);
        return true;
    }

    return false;
}

bool WorksModel::removeItem(const QString &workId)
{
    if(workId.isEmpty())
        return false;

    for(int index = 0; index < m_items.size(); ++index)
    {
        if(m_items.at(index).id != workId)
            continue;

        beginRemoveRows(QModelIndex(), index, index);
        m_items.removeAt(index);
        endRemoveRows();
        emit sizeChanged();
        return true;
    }

    return false;
}

QVariantMap WorksModel::itemData(const QString &workId) const
{
    for(const WorkItem &item : getItems())
    {
        if(item.id == workId)
            return item.toMap();
    }
    return {};
}

QVariantList WorksModel::items() const
{
    return variantListFromItems(getItems());
}

} // namespace workorder
